### ISSUE TITLE
Please add/fix support for assume role CLI profiles in .aws/credentials

### ISSUE THREAD

**Issue by et304383** [aws/aws-sdk-php#1323]
#1030 was closed and appears to remain unresolved as of 3.31.0.

I expect that if I have this in `.aws/credentials`:

```
[profile1]
output = json
region = us-east-1
aws_access_key_id = A**************
aws_secret_access_key = *****************

[profile2]
output = json
region = us-east-1
role_arn = arn:aws:iam::123456789012:role/Developer
source_profile = profile1
```

I should be able to simply set the env var:

```
export AWS_PROFILE=profile2
```

And then this should work:
```
$ec2_client = new Aws\Ec2\Ec2Client([
  'region' => 'us-east-1',
  'version' => 'latest'
]);
```

This does not.  As I noted on the other issue, CredentialsProvider is throwing this error:

```
Fatal error: Uncaught Aws\Exception\CredentialsException: No credentials present in INI profile 'profile2' (/var/www/.aws/credentials) in /var/www/html/vendor/aws/aws-sdk-php/src/Credentials/CredentialProvider.php:394
```

This should work exactly like boto3 and just work.  The SDK should detect that the profile is an assume role type with a role_arn value and perform the assume role call to get credentials.  **I should not have to modify the code to explicitly declare a credentials provider**.

If I am simply missing a step please clarify, but I believe code should be agnostic to whether it's running with the env var AWS_PROFILE set or running on EC2.

**Comment by imshashank** [comment]
@eric-tucker Sorry, that you are facing this issue. Please give us some time to repro this issue.

**Comment by eric-tucker** [comment]
Thanks!  Note I fixed my source_profile value (was profile1).  That was just a typo in my example.  My real config is correct.  :)

**Comment by imshashank** [comment]
@eric-tucker Hi, I tested with setting  `AWS_PROFILE` to point to the different account and it worked for me.  Environment variables set in `bash` using `export` for a terminal are only valid for that session until the terminal is closed.

The php script must be invoked inside the same terminal or it won't be able to access the env variable. You can use `getenv['AWS_PROFILE']` to check what is the value of env variable that is being seen in your php script.

**Comment by eric-tucker** [comment]
So to start, I'm not using a terminal.  I'm doing this through a web server.  But to prove this isn't working on the CLI either, here are my files:
```
[etucker: php_profiles_test]$ ll
total 36
-rw-rw-r-- 1 etucker etucker    54 Jul  4 11:19 composer.json
-rw-rw-r-- 1 etucker etucker 12972 Jul 12 14:03 composer.lock
-rw-rw-r-- 1 etucker etucker   236 Jul 12 14:02 index.php
drwxrwxr-x 8 etucker etucker  4096 Jul 12 14:03 vendor
```

composer.json:
```
{
  "require": {
    "aws/aws-sdk-php": "3.31.0"
  }
}
```

index.php:
```
<?php
var_dump(getenv('HOME'));
var_dump(getenv('AWS_PROFILE'));

require 'vendor/autoload.php';

$ec2_client = new Aws\Ec2\Ec2Client([
  'region' => 'us-east-1',
  'version' => 'latest'
]);

var_dump($ec2_client->describeInstances());
```

Running:
```
[etucker: php_profiles_test]$ export AWS_PROFILE=profile2
[etucker: php_profiles_test]$ php index.php 
/home/etucker/play/php_profiles_test/index.php:2:
string(13) "/home/etucker"
/home/etucker/play/php_profiles_test/index.php:3:
string(4) "profile2"
PHP Fatal error:  Uncaught Aws\Exception\CredentialsException: Error retrieving credentials from the instance profile metadata server. (cURL error 28: Connection timed out after 1001 milliseconds (see http://curl.haxx.se/libcurl/c/libcurl-errors.html)) in /home/etucker/play/php_profiles_test/vendor/aws/aws-sdk-php/src/Credentials/InstanceProfileProvider.php:79
Stack trace:
#0 /home/etucker/play/php_profiles_test/vendor/guzzlehttp/promises/src/Promise.php(203): Aws\Credentials\InstanceProfileProvider->Aws\Credentials\{closure}(Object(GuzzleHttp\Exception\ConnectException))
#1 /home/etucker/play/php_profiles_test/vendor/guzzlehttp/promises/src/Promise.php(156): GuzzleHttp\Promise\Promise::callHandler(2, Array, Array)
#2 /home/etucker/play/php_profiles_test/vendor/guzzlehttp/promises/src/TaskQueue.php(47): GuzzleHttp\Promise\Promise::GuzzleHttp\Promise\{closure}()
#3 /home/etucker/play/php_profiles_test/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php(96): GuzzleHttp\Promise\TaskQueue->run()
#4 /home/etucker/play/php_pro in /home/etucker/play/php_profiles_test/vendor/aws/aws-sdk-php/src/Credentials/InstanceProfileProvider.php on line 79
[etucker: php_profiles_test]$ 
```

So the profile isn't even working.  It's trying to call out to the metadata server.

If I explicitly set the profile:
index.php
```
<?php
var_dump(getenv('HOME'));
var_dump(getenv('AWS_PROFILE'));

require 'vendor/autoload.php';

$ec2_client = new Aws\Ec2\Ec2Client([
  'profile' => 'profile2',
  'region' => 'us-east-1',
  'version' => 'latest'
]);

var_dump($ec2_client->describeInstances());
```
Running:
```
[etucker: php_profiles_test]$ php index.php 
/home/etucker/play/php_profiles_test/index.php:2:
string(13) "/home/etucker"
/home/etucker/play/php_profiles_test/index.php:3:
string(4) "profile2"
PHP Fatal error:  Uncaught Aws\Exception\CredentialsException: No credentials present in INI profile 'profile2' (/home/etucker/.aws/credentials) in /home/etucker/play/php_profiles_test/vendor/aws/aws-sdk-php/src/Credentials/CredentialProvider.php:394
Stack trace:
#0 /home/etucker/play/php_profiles_test/vendor/aws/aws-sdk-php/src/Credentials/CredentialProvider.php(303): Aws\Credentials\CredentialProvider::reject('No credentials ...')
#1 /home/etucker/play/php_profiles_test/vendor/aws/aws-sdk-php/src/Middleware.php(122): Aws\Credentials\CredentialProvider::Aws\Credentials\{closure}()
#2 /home/etucker/play/php_profiles_test/vendor/aws/aws-sdk-php/src/RetryMiddleware.php(159): Aws\Middleware::Aws\{closure}(Object(Aws\Command), Object(GuzzleHttp\Psr7\Request))
#3 /home/etucker/play/php_profiles_test/vendor/aws/aws-sdk-php/src/Middleware.php(207): Aws\RetryMiddleware->__invoke(Object(Aws\Command), Object(GuzzleHttp\Psr7\Request))
#4 /home/etucker/play/php_profiles_test/vendor/aws/aws-sdk-php/src/ClientResolver.php(582): Aws\Middleware:: in /home/etucker/play/php_profiles_test/vendor/aws/aws-sdk-php/src/Credentials/CredentialProvider.php on line 394
```

So in this case I see the explicit error that the credentials aren't set.

Put it this way - tell me where in the CredentialsProvider logic that you're checking for role_arn?

I grepped the source code for `role_arn` and got no results.

```
[etucker: aws-sdk-php]$ grep -r 'role_arn' .
[etucker: aws-sdk-php]$  
```

There is no logic to support role based profile in the source code.

**Comment by imshashank** [comment]
@eric-tucker I am sorry I misunderstood. PHP SDK doesn't support assuming role using the credentials file. You will have to explicitly tell the SDK to assumeRole.

You can see this example on how to use STS client for assumeRole-> http://docs.aws.amazon.com/aws-sdk-php/v3/guide/guide/credentials.html#using-assume-role-credentials

Thanks for notifying this feature to us and we will add this to our backlog to add support for assume-role based on `role_arn` in credentials file.

**Comment by eric-tucker** [comment]
Thank you for clarifying / admitting. That is exactly why I created #1030 a year ago.  It was closed but still hasn't been resolved.

**Comment by lifeofguenter** [comment]
+1

**Comment by expertcoder** [comment]
It would also be nice if you could assume the role via the constructor argument also as well as `AWS_PROFILE` environmental variable

```
$s3 = new Aws\S3\S3Client([
    'region' => 'us-east-1',
    'profile' => 'myrole',
    'version' => 'latest',
]);
```
This is not supported either it seems.

**Comment by jcolfej** [comment]
This bug is very annoying and prevents to develop properly ...
Could we fix it definitively ? Knowing that it is a major "bug" !

**Comment by expertcoder** [comment]
If you have time and energy, you an write your own "Credential Provider" in PHP to do almost anything you want (https://docs.aws.amazon.com/sdk-for-php/v3/developer-guide/guide_credentials.html). This happened to suit my needs, but understand that its annoying to have to spend time to implement something which could be supported by default.

**Comment by jcolfej** [comment]
The problem with that is that you must write 2 authentifications method : 

- one for developement with own credential provider
- one for production that use EC2 Env to auth

And you must detect if you are in production or not ... And this is not testable ...

The problem is that the SDK don't know what is "source_profile" ...

**Comment by PavloPixart** [comment]
Hi there.
Is there any news on that topic? I think that this topic could be related to: https://github.com/aws/aws-sdk-php/issues/1030 
Is there any idea on how much it will take to be fixed? 
Thank you.

**Comment by et304383** [comment]
@PavloPixart of course it's related to 1030.  That's why I mentioned it in the first line of the issue and again later on.

**Comment by PavloPixart** [comment]
@et304383 Apologies for that, I was focused on searching any answer on the latest messages. I saw that there have been no answer since 28 May. Do you think there is any chance to escalate to someone that could fix this problem sooner than later?

**Comment by expertcoder** [comment]
I solved this problem months ago, I cant remember the specifics very well.

Here is how I solved this with Symfony 4, I will cut and paste my Symfony config as inspiration. I believe you should be able to manually instantiate classes in pure PHP code if you are not using Symfony, and just try to replicate what the config files are doing. Sorry I can't be of better help but it would take me a while to remember the exact steps and provide a generic PHP example.

**CredentialProvider.yml** (For prod/EC2) from memory this will just use the attached role

```
services:
  aws_credential_provider:
    class: Aws\Credentials\InstanceProfileProvider
    public: false
    factory: ['Aws\Credentials\CredentialProvider', instanceProfile]
```

**CredentialProvider_dev.yml** (For "local" environments, to use a profile in `~/.aws/credentials` folder)

```
# For the dev environment we first use the AWS Credentials from our AWS CLI profile
# located in `~/.aws/credentials`. However PHP don't use the credentials directly, rather
# we use them to assume an AWS Role, and PHP will use that role to make API calls.
# `aws_sts_client` service is only needed on dev for assuming the role

services:
  aws_credential_provider:
    class: Aws\Credentials\AssumeRoleCredentialProvider
    public: false
    factory: ['Aws\Credentials\CredentialProvider', assumeRole]
    arguments:
      - client: '@aws_sts_client'
        assume_role_params:
            RoleArn: '%aws_role_arn%'
            RoleSessionName: 'my_php_session' # This is an arbitrary name
        version: latest

  aws_sts_client:
    class: Aws\Sts\StsClient
    public: false
    arguments:
      - profile: '%aws_profile%'
        region: '%aws_region%'
        version: latest
```
**S3Client.yml**

```
services:
  Aws\S3\S3Client:
    public: true
    arguments:
      - credentials: '@aws_credential_provider'
        region: '%aws_region%'
        version: latest
```
**Usage:**

```
$s3Client = $this->container->get('Aws\S3\S3Client')
```

**Comment by PavloPixart** [comment]
@expertcoder : Thank you for your solutions and I think it could be a possible way to solve the problem. But I think I will wait.

I think that the official SDK should be fixed because the way I was using it was the proper way. I hope that someone that is working on this SDK will be able to fix the problem with an official release.

**Comment by PavloPixart** [comment]
Any update?

**Comment by howardlopez** [comment]
I can't give any set timeline, but this is in our backlog and we are definitely paying attention to this feedback.

**Comment by PavloPixart** [comment]
I think that a minimum deadline should be set. To get our development stage working we have assign all the permission to a non admin account. So the current status isn't secure in any way. Is it possible to have an estimation on the time you will need to fix the bug? Thank you.

**Comment by PavloPixart** [comment]
Any update?

**Comment by sublimited** [comment]
is there a way to use arn,acc,userid from `aws sts get-caller-identity` to request firehose, without having to put the key and secret into credentials? just like the kinesis CLI tool does?
Thanks

**Comment by PavloPixart** [comment]
A month has passed till the last update request. We are still working with an unsafe way of managing credentials in our development environment. Is there any way to get at least the required time to fix it as well as an estimate date for the release? Thank you.

**Comment by PavloPixart** [comment]
Any update? Another month has passed more or less.

**Comment by jameshulse** [comment]
I'm running into this issue too and will be a show stopper for using the aws php sdk.

**Comment by PavloPixart** [comment]
I've lost any hope on the resolution of this issue. It seems strange that a topic like that could be ignored this way. We are generating credentials as root accounts that means that we can compromise our enviroments also from our local machine.

**Comment by srchase** [comment]
Resolved in #1804.

It will go out with the next release of the SDK.
